src/Course/StateT.hs

```diff
@@ -164,25 +114,33 @@ getT =
 --
 -- >>> runStateT (putT 2 :: StateT Int List ()) 0
 -- [((),2)]
-putT ::
-  Applicative f =>
-  s
-  -> StateT s f ()
-putT =
-  error "todo: Course.StateT#putT"
+putT :: Applicative f => s -> StateT s f ()
+putT s = StateT (\_ -> pure ((), s))
 
 -- | Remove all duplicate elements in a `List`.
 --
 -- /Tip:/ Use `filtering` and `State'` with a @Data.Set#Set@.
 --
 -- prop> \xs -> distinct' xs == distinct' (flatMap (\x -> x :. x :. Nil) xs)
-distinct' ::
-  (Ord a, Num a) =>
-  List a
-  -> List a
-distinct' =
-  error "todo: Course.StateT#distinct'"
+distinct' :: (Ord a, Num a) => List a -> List a
+distinct' xs = eval' (filtering p xs) S.empty
+  where
+    p x = StateT (\s -> ExactlyOne (S.notMember x s, S.insert x s))
```

Comment:
it's funny... I feel like this pattern you're both using is less clear than using the monadic interface, for some reason. I'm not entirely sure why. It definitely makes things neater and leads to better solutions sometimes, tho, so it's cool :)

Reply:
I like yours, I want to take the time to understand it, I guess I'll have it now :)